### PR TITLE
feat: flag to manage sequenced rollout #201

### PR DESCRIPTION
**Pull request by srinandan**

### CODE REVIEW

**Review by ssvaidyanathan — APPROVED** [review]

**Comment by ssvaidyanathan** [comment]
Still unclear how this field will fix the issue? I called the Mgmt API with that property set to `true` and was able to deploy the same revision to two environments part of the same env group. So how will the routing happen? Will this field make sure that the call goes to the latest deployed environment?

**Comment by kurtkanaskie** [comment]
There's an error on the query param for the flag:
has: q.Set("sequencedRollout", serviceAccountName)
should be: q.Set("sequencedRollout", sequencedRollout)

**Comment by kurtkanaskie** [comment]
Just testing this out now, the use of sequencedRollout true or false does not prevent the deployment to another env in the same envGroup.

The use of generateDeployChangeReport will indicate: "New deployment will not receive any traffic for basepath /notarget, as this basepath is already claimed by proxy notarget in environment test."

It's up to the user in the UI or the tool invoking the APIs to choose to continue with deployment or not. Even in the UI I can ignore the generateDeployChangeReport and deploy to another env in the same envGroup.

File: cmd/apis/depapi.go [hunk 1]
```diff
@@ -100,7 +100,9 @@ func init() {
 	DepCmd.Flags().BoolVarP(&wait, "wait", "",
 		false, "Waits for the deployment to finish, with success or error")
 	DepCmd.Flags().BoolVarP(&sequencedRollout, "sequencedRollout", "",
-		false, "If set to true, the routing rules will be rolled out in a safe order; default is false")
+		true, "If set to true, the routing rules will be rolled out in a safe order; default is true")
+	DepCmd.Flags().BoolVarP(&safeDeploy, "safedeploy", "",
+		false, "If set to true, generateDeployChangeReport will executed and deployment will proceed if there are no conflicts; default is false")
```
Comment (ssvaidyanathan) [inline cmd/apis/depapi.go]: will `be` executed

**Comment by ssvaidyanathan** [comment]
> Why is `generateDeployChangeReport` default false? Shouldn't it be true to prevent unexpected behavior, like having a proxy deployed to 2 envs in the same envGroup?

Probably for backward compatibility?

**Comment by kurtkanaskie** [comment]
> Probably for backward compatibility?

I think that behavior is a bug.

**Review by ssvaidyanathan — APPROVED** [review]

File: cmd/apis/undepapi.go [hunk 2]
```diff
@@ -37,18 +37,23 @@ var UndepCmd = &cobra.Command{
...
 func init() {
 	UndepCmd.Flags().StringVarP(&name, "name", "n",
 		"", "API proxy name")
 	UndepCmd.Flags().StringVarP(&env, "env", "e",
 		"", "Apigee environment name")
 	UndepCmd.Flags().IntVarP(&revision, "rev", "v",
 		-1, "API Proxy revision")
+	UndepCmd.Flags().BoolVarP(&safeUndeploy, "safeundeploy", "",
+		true, "When set to true, generateUndeployChangeReport will be executed and "+
+			"deployment will proceed if there are no conflicts; default is true")
```
Comment (kurtkanaskie) [inline cmd/apis/undepapi.go:56]: undeployment will proceed

**Review by kurtkanaskie — CHANGES_REQUESTED** [review]
One minor editorial on L56

undeployment will proceed...

**Review by kurtkanaskie — APPROVED** [review]

**Comment by Pavloro** [comment]
It works as expected now. Thank you for your quick response and fix!

**Pull request merged**